### ISSUE TITLE
[rllib] Displaying evaluation metrics on tensorboard

### ISSUE THREAD

**Issue by jangkj09** [ray-project/ray#8429]
### What is your question?
Hi, this is a simple question and I'm not sure what is wrong. I am testing out an evaluation loop during training and I can see the metrics output on the terminal, but I am unable to have them displayed on tensorboard. I have some custom metrics which I implemented according to the example and I can see those on tensorboard. From the documentation, I thought that evaluation metrics would show up automatically on tensorboard. Is this correct? Or do I need to make an adjustment?

*Ray version and other system information (Python version, TensorFlow version, OS):*
Python 3.6
ray == 0.8.5
tensorflow == 2.1.0
Ubuntu 18.04

**Comment by ericl** [comment]
Yes, evaluation metric should show up under the evaluation/ key.

**Comment by jangkj09** [comment]
Then is there a way to debug why it isn't showing up? I didn't declare a custom evaluation function or metric and simply set values for evaluation_interval config, evaluation_num_episodes,  and evaluation_num_workers. The values print out in the console, but they are missing from tensorboard. I saw a different post and I thought that it may be an issue with tensorboard and so I uninstalled tensorboard and left only tensorboardX, but there was no difference.  Thanks.

**Comment by ericl** [comment]
Try running `rllib train --run=PG --env=CartPole-v0 -v --config='{"evaluation_num_workers": 2, "evaluation_interval": 1}'`, you should see:

![image](https://user-images.githubusercontent.com/14922/82001621-6cf72200-9610-11ea-9a78-beec651a519c.png)

**Comment by jangkj09** [comment]
Yes, I do see that as well. This is a snippet of the code I am running and how I configure the training and configuring evaluation. It is working because I do see the evaluation output, but it doesn't show up in tensorboard. The only difference I can imagine is either because I am using a custom environment and I also have some custom metrics (defined and configured like the example code). Or does this have to do with the way I am interacting with the training instance?

```
config["callbacks"] = MyCallbacks
#Evaluation config
config["evaluation_num_workers"]=2
config["evaluation_interval"]=1
# Run 100 episodes each time evaluation runs.
config["evaluation_num_episodes"]=100

trainer = ppo.PPOTrainer(config=config,
                         env="custom_env-v0")

evaluation_results=[]
for i in range(1000):
   # Perform one iteration of training the policy with PPO
   result = trainer.train()
   #print(pretty_print(result))
   print(evaluation_results)
   if i % 100 == 0:
       checkpoint = trainer.save()
       print("checkpoint saved at", checkpoint)
```

**Comment by jpmediadev** [comment]
Hi, same problem 
i have a Custom Eval Function  config['custom_eval_function'] = CustomTester().run
which returns {'evaluation_rewards': 14.342403665942745}
i can see this in result = trainer.train()
but nothing in Tensorboard

**Comment by cassidylaidlaw** [comment]
I'm getting the same issue. It looks like it's caused by how `Trainable.train` and `Trainer.train` interact. In `Trainable.train`, the control flow goes

    ...
    self._log_result(result)
    return result

And then in `Trainer.train`, the control flow

    result = Trainable.train(self)
    ...
    evaluation_metrics = self._evaluate()
    result.update(evaluation_metrics)
    return result

So evaluation metrics are getting merged into the result after it has already been logged. Not sure how to fix this without some restructuring.

**Comment by JenishPatel99** [comment]
@cassidylaidlaw We have been getting the same issue with our custom evaluation metrics not being shown on tensorboard. Our method to invoke the train method is as followed:

trainer = ppo.PPOTrainer(config=config, env=env)

**Comment by dimzhead** [comment]
pip install tensorboardX 

This did the trick for me

**Comment by drmeerkat** [comment]
I have to say, till 2024, it still feels like a mystery to me how to log information in tensorboard while using rllib...It might be inevitable when the software package is highly integrated.
